PD placement rules should be enabled if using TiFlash

## Bug Report

**What version of TiDB Operator are you using?**
<!-- You can run `kubectl exec -n tidb-admin {tidb-controller-manager-pod-name} -- tidb-controller-manager -V` -->
latest release

**What did you do?**
<!-- If possible, provide a recipe for reproducing the error. How you installed tidb-operator and tidb-cluster. -->

Start TiDBCluster with TiFlash

**What did you expect to see?**

TiFlash and TiDBcluster running normally

**What did you see instead?**

TiFlash failed to start because the PD placement-rules is not enabled, this is a prerequisite for TiFlash. Error log of TiFlash

```
klo yeya24-tiflash-0 -c proxylog
[2020/04/16 19:27:56.857 +00:00] [ERROR] [util.rs:327] ["request failed"] [err="Grpc(RpcFailure(RpcStatus { status: FailedPrecondition, details: Some(\"placement rules is disabled\") }))"]
[2020/04/16 19:27:56.857 +00:00] [ERROR] [util.rs:327] ["request failed"] [err="Grpc(RpcFailure(RpcStatus { status: FailedPrecondition, details: Some(\"placement rules is disabled\") }))"]
[2020/04/16 19:27:56.857 +00:00] [FATAL] [server.rs:287] ["failed to start node: Other(\"[src/pd/util.rs:335]: fail to request\")"]
```

PD config status:

```
pdctl config show | grep enable-placement
    "enable-placement-rules": "false",
```

If I manually enabled replacement rules in PD, then TiFlash starts perfectly

--- a/pkg/manager/member/pd_member_manager.go
+++ b/pkg/manager/member/pd_member_manager.go
@@ -741,17 +741,7 @@ func getPDConfigMap(tc *v1alpha1.TidbCluster) (*corev1.ConfigMap, error) {
 		config.Dashboard.TiDBCertPath = path.Join(tidbClientCertPath, corev1.TLSCertKey)
 		config.Dashboard.TiDBKeyPath = path.Join(tidbClientCertPath, corev1.TLSPrivateKeyKey)
 	}
-	// TiFlash requires PD to enable the `replication.enable-placement-rules`
-	// Check detail in https://pingcap.com/docs/stable/reference/tiflash/deploy/
-	if tc.Spec.TiFlash != nil {
-		if config.Replication == nil {
-			config.Replication = &v1alpha1.PDReplicationConfig{}
-		}
-		if config.Replication.EnablePlacementRules == nil {
-			enable := true
-			config.Replication.EnablePlacementRules = &enable
-		}
-	}
+
 	confText, err := MarshalTOML(config)
 	if err != nil {
 		return nil, err

--- a/pkg/manager/member/tiflash_member_manager.go
+++ b/pkg/manager/member/tiflash_member_manager.go
@@ -107,12 +107,34 @@ func (tfmm *tiflashMemberManager) Sync(tc *v1alpha1.TidbCluster) error {
 		return controller.RequeueErrorf("TidbCluster: [%s/%s], waiting for PD cluster running", ns, tcName)
 	}
 
+	err := tfmm.enablePlacementRules(tc)
+	if err != nil {
+		klog.Errorf("Enable placement rules failed, error: %v", err)
+		// No need to return err here, just continue to sync tiflash
+	}
 	// Sync TiFlash Headless Service
-	if err := tfmm.syncHeadlessService(tc); err != nil {
+	if err = tfmm.syncHeadlessService(tc); err != nil {
 		return err
 	}
 
 	return tfmm.syncStatefulSet(tc)
+}
+
+func (tfmm *tiflashMemberManager) enablePlacementRules(tc *v1alpha1.TidbCluster) error {
+	pdCli := controller.GetPDClient(tfmm.pdControl, tc)
+	config, err := pdCli.GetConfig()
+	if err != nil {
+		return err
+	}
+	if config.Replication.EnablePlacementRules != nil && (!*config.Replication.EnablePlacementRules) {
+		klog.Infof("Cluster %s/%s enable-placement-rules is %v, set it to true", tc.Namespace, tc.Name, *config.Replication.EnablePlacementRules)
+		enable := true
+		rule := pdapi.PDReplicationConfig{
+			EnablePlacementRules: &enable,
+		}
+		return pdCli.UpdateEnablePlacementRules(rule)
+	}
+	return nil
 }
 
 func (tfmm *tiflashMemberManager) syncHeadlessService(tc *v1alpha1.TidbCluster) error {

--- a/pkg/manager/member/tiflash_member_manager.go
+++ b/pkg/manager/member/tiflash_member_manager.go
@@ -129,10 +129,10 @@ func (tfmm *tiflashMemberManager) enablePlacementRules(tc *v1alpha1.TidbCluster)
 	if config.Replication.EnablePlacementRules != nil && (!*config.Replication.EnablePlacementRules) {
 		klog.Infof("Cluster %s/%s enable-placement-rules is %v, set it to true", tc.Namespace, tc.Name, *config.Replication.EnablePlacementRules)
 		enable := true
-		rule := pdapi.PDReplicationConfig{
+		rep := pdapi.PDReplicationConfig{
 			EnablePlacementRules: &enable,
 		}
-		return pdCli.UpdateEnablePlacementRules(rule)
+		return pdCli.UpdateReplicationConfig(rep)
 	}
 	return nil
 }

--- a/pkg/pdapi/pdapi.go
+++ b/pkg/pdapi/pdapi.go
@@ -154,8 +154,8 @@ type PDClient interface {
 	// storeLabelsEqualNodeLabels compares store labels with node labels
 	// for historic reasons, PD stores TiKV labels as []*StoreLabel which is a key-value pair slice
 	SetStoreLabels(storeID uint64, labels map[string]string) (bool, error)
-	// UpdateEnablePlacementRules updates the enable-placement-rules config
-	UpdateEnablePlacementRules(rule PDReplicationConfig) error
+	// UpdateReplicationConfig updates the replication config
+	UpdateReplicationConfig(config PDReplicationConfig) error
 	// DeleteStore deletes a TiKV store from cluster
 	DeleteStore(storeID uint64) error
 	// SetStoreState sets store to specified state.
@@ -515,9 +515,9 @@ func (pc *pdClient) SetStoreLabels(storeID uint64, labels map[string]string) (bo
 	return false, fmt.Errorf("failed %v to set store labels: %v", res.StatusCode, err2)
 }
 
-func (pc *pdClient) UpdateEnablePlacementRules(rule PDReplicationConfig) error {
+func (pc *pdClient) UpdateReplicationConfig(config PDReplicationConfig) error {
 	apiURL := fmt.Sprintf("%s/%s", pc.url, pdReplicationPrefix)
-	data, err := json.Marshal(rule)
+	data, err := json.Marshal(config)
 	if err != nil {
 		return err
 	}
@@ -529,8 +529,8 @@ func (pc *pdClient) UpdateEnablePlacementRules(rule PDReplicationConfig) error {
 	if res.StatusCode == http.StatusOK {
 		return nil
 	}
-	err2 := httputil.ReadErrorBody(res.Body)
-	return fmt.Errorf("failed %v to update enable-placement-rules: %v", res.StatusCode, err2)
+	err = httputil.ReadErrorBody(res.Body)
+	return fmt.Errorf("failed %v to update replication: %v", res.StatusCode, err)
 }
 
 func (pc *pdClient) BeginEvictLeader(storeID uint64) error {
@@ -718,7 +718,7 @@ const (
 	DeleteMemberByIDActionType         ActionType = "DeleteMemberByID"
 	DeleteMemberActionType             ActionType = "DeleteMember "
 	SetStoreLabelsActionType           ActionType = "SetStoreLabels"
-	UpdatePlacementRulesActionType     ActionType = "UpdateEnablePlacementRules"
+	UpdateReplicationActionType        ActionType = "UpdateReplicationConfig"
 	BeginEvictLeaderActionType         ActionType = "BeginEvictLeader"
 	EndEvictLeaderActionType           ActionType = "EndEvictLeader"
 	GetEvictLeaderSchedulersActionType ActionType = "GetEvictLeaderSchedulers"
@@ -735,10 +735,10 @@ func (nfr *NotFoundReaction) Error() string {
 }
 
 type Action struct {
-	ID     uint64
-	Name   string
-	Labels map[string]string
-	Rule   PDReplicationConfig
+	ID          uint64
+	Name        string
+	Labels      map[string]string
+	Replication PDReplicationConfig
 }
 
 type Reaction func(action *Action) (interface{}, error)
@@ -878,10 +878,10 @@ func (pc *FakePDClient) SetStoreLabels(storeID uint64, labels map[string]string)
 	return true, nil
 }
 
-// UpdateEnablePlacementRules updates the enable-placement-rules
-func (pc *FakePDClient) UpdateEnablePlacementRules(rule PDReplicationConfig) error {
-	if reaction, ok := pc.reactions[UpdatePlacementRulesActionType]; ok {
-		action := &Action{Rule: rule}
+// UpdateReplicationConfig updates the replication config
+func (pc *FakePDClient) UpdateReplicationConfig(config PDReplicationConfig) error {
+	if reaction, ok := pc.reactions[UpdateReplicationActionType]; ok {
+		action := &Action{Replication: config}
 		_, err := reaction(action)
 		return err
 	}

--- a/pkg/pdapi/pdapi.go
+++ b/pkg/pdapi/pdapi.go
@@ -154,6 +154,8 @@ type PDClient interface {
 	// storeLabelsEqualNodeLabels compares store labels with node labels
 	// for historic reasons, PD stores TiKV labels as []*StoreLabel which is a key-value pair slice
 	SetStoreLabels(storeID uint64, labels map[string]string) (bool, error)
+	// UpdateEnablePlacementRules updates the enable-placement-rules config
+	UpdateEnablePlacementRules(rule PDReplicationConfig) error
 	// DeleteStore deletes a TiKV store from cluster
 	DeleteStore(storeID uint64) error
 	// SetStoreState sets store to specified state.
@@ -185,6 +187,7 @@ var (
 	schedulersPrefix       = "pd/api/v1/schedulers"
 	pdLeaderPrefix         = "pd/api/v1/leader"
 	pdLeaderTransferPrefix = "pd/api/v1/leader/transfer"
+	pdReplicationPrefix    = "pd/api/v1/config/replicate"
 )
 
 // pdClient is default implementation of PDClient
@@ -512,6 +515,24 @@ func (pc *pdClient) SetStoreLabels(storeID uint64, labels map[string]string) (bo
 	return false, fmt.Errorf("failed %v to set store labels: %v", res.StatusCode, err2)
 }
 
+func (pc *pdClient) UpdateEnablePlacementRules(rule PDReplicationConfig) error {
+	apiURL := fmt.Sprintf("%s/%s", pc.url, pdReplicationPrefix)
+	data, err := json.Marshal(rule)
+	if err != nil {
+		return err
+	}
+	res, err := pc.httpClient.Post(apiURL, "application/json", bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+	defer httputil.DeferClose(res.Body)
+	if res.StatusCode == http.StatusOK {
+		return nil
+	}
+	err2 := httputil.ReadErrorBody(res.Body)
+	return fmt.Errorf("failed %v to update enable-placement-rules: %v", res.StatusCode, err2)
+}
+
 func (pc *pdClient) BeginEvictLeader(storeID uint64) error {
 	leaderEvictInfo := getLeaderEvictSchedulerInfo(storeID)
 	apiURL := fmt.Sprintf("%s/%s", pc.url, schedulersPrefix)
@@ -697,6 +718,7 @@ const (
 	DeleteMemberByIDActionType         ActionType = "DeleteMemberByID"
 	DeleteMemberActionType             ActionType = "DeleteMember "
 	SetStoreLabelsActionType           ActionType = "SetStoreLabels"
+	UpdatePlacementRulesActionType     ActionType = "UpdateEnablePlacementRules"
 	BeginEvictLeaderActionType         ActionType = "BeginEvictLeader"
 	EndEvictLeaderActionType           ActionType = "EndEvictLeader"
 	GetEvictLeaderSchedulersActionType ActionType = "GetEvictLeaderSchedulers"
@@ -716,6 +738,7 @@ type Action struct {
 	ID     uint64
 	Name   string
 	Labels map[string]string
+	Rule   PDReplicationConfig
 }
 
 type Reaction func(action *Action) (interface{}, error)
@@ -853,6 +876,16 @@ func (pc *FakePDClient) SetStoreLabels(storeID uint64, labels map[string]string)
 		return result.(bool), err
 	}
 	return true, nil
+}
+
+// UpdateEnablePlacementRules updates the enable-placement-rules
+func (pc *FakePDClient) UpdateEnablePlacementRules(rule PDReplicationConfig) error {
+	if reaction, ok := pc.reactions[UpdatePlacementRulesActionType]; ok {
+		action := &Action{Rule: rule}
+		_, err := reaction(action)
+		return err
+	}
+	return nil
 }
 
 func (pc *FakePDClient) BeginEvictLeader(storeID uint64) error {

